src/NodaTime/Xml/XmlSchemaDefinition.cs

```diff
@@ -0,0 +1,238 @@
+            YearMonthSchemaType = CreatePatternRestriction<YearMonth>(xsStringType, $"{YearPattern}-{MonthPattern}");
+            ZonedDateTimeSchemaType = CreateSchemaTypeWithAttributes<ZonedDateTime>(OffsetDateTimeSchemaType, zoneAttribute, calendarAttribute);
+
+            DependentSchemaTypes = new Dictionary<XmlSchemaType, IEnumerable<XmlSchemaType>>
+            {
+                [IntervalSchemaType] = new[] { InstantSchemaType },
+                [LocalDateSchemaType] = new[] { localDateRestriction, calendarRestriction },
+                [LocalDateTimeSchemaType] = new[] { localDateTimeRestriction, calendarRestriction },
+                [ZonedDateTimeSchemaType] = new[] { OffsetDateTimeSchemaType, calendarRestriction, zoneRestriction },
+            };
+
+            NodaTimeXmlSchema = CreateNodaTimeXmlSchema();
+        }
+
+        private static readonly XmlSchemaType AnnualDateSchemaType;
+        private static readonly XmlSchemaType DurationSchemaType;
+        private static readonly XmlSchemaType InstantSchemaType;
+        private static readonly XmlSchemaType IntervalSchemaType;
+        private static readonly XmlSchemaType LocalDateSchemaType;
+        private static readonly XmlSchemaType LocalDateTimeSchemaType;
+        private static readonly XmlSchemaType LocalTimeSchemaType;
+        private static readonly XmlSchemaType OffsetSchemaType;
+        private static readonly XmlSchemaType OffsetDateSchemaType;
+        private static readonly XmlSchemaType OffsetDateTimeSchemaType;
+        private static readonly XmlSchemaType OffsetTimeSchemaType;
+        private static readonly XmlSchemaType PeriodBuilderSchemaType;
+        private static readonly XmlSchemaType YearMonthSchemaType;
+        private static readonly XmlSchemaType ZonedDateTimeSchemaType;
+        private static readonly Dictionary<XmlSchemaType, IEnumerable<XmlSchemaType>> DependentSchemaTypes;
+
+        private static XmlQualifiedName AddSchemaType(this XmlSchemaSet schemaSet, XmlSchemaType schemaType)
+        {
+            var schema = new XmlSchema
+            {
+                TargetNamespace = NodaTimeXmlNamespace.Namespace,
+                Items = { schemaType }
+            };
+            if (DependentSchemaTypes.TryGetValue(schemaType, out var dependentSchemaTypes))
+            {
+                foreach (var dependentSchemaType in dependentSchemaTypes)
+                {
+                    schema.Items.Add(dependentSchemaType);
+                }
+            }
+            schemaSet.Add(schema);
+            return schemaType.QualifiedName;
+        }
+
+        private static XmlSchema CreateNodaTimeXmlSchema()
+        {
+            var schemaSetForCollecting = new XmlSchemaSet();
+            var xmlSerializableTypes = typeof(XmlSchemaDefinition).Assembly.GetExportedTypes().Where(t => t.GetInterfaces().Contains(typeof(IXmlSerializable)));
```

Comment:
Rather than doing all of this via reflection, how about we have a hard-coded list of serializable types, and then a unit test that *does* use reflection to make sure it's accurate?

Reply:
Absolutely, done in 6c06ed466ce2770bc44e2b7d4e0199b807fd2c7d.